crates/llm-local/src/llama.rs

```diff
@@ -0,0 +1,168 @@
+use crate::{token_output_stream, utils::load_safetensors, CachedInferencingModel};
+use anyhow::{anyhow, Result};
+use candle::{utils, Device, Tensor};
+use candle_nn::VarBuilder;
+use candle_transformers::{
+    generation::{LogitsProcessor, Sampling},
+    models::llama::{self, Cache, Config, Llama, LlamaConfig},
+};
+use rand::{RngCore, SeedableRng};
+use spin_core::async_trait;
+use spin_world::v2::llm::{self as wasi_llm, InferencingUsage};
+use std::{fs, path::Path, sync::Arc};
+use tokenizers::Tokenizer;
+
+const TOKENIZER_FILENAME: &str = "tokenizer.json";
+const CONFIG_FILENAME: &str = "config.json";
+const EOS_TOKEN: &str = "</s>";
+const MODEL_SAFETENSORS_INDEX_FILE: &str = "model.safetensors.index.json";
+
+pub fn auto_device() -> Result<Device> {
+    if utils::cuda_is_available() {
+        Ok(Device::new_cuda(0)?)
+    } else if utils::metal_is_available() {
+        Ok(Device::new_metal(0)?)
+    } else {
+        Ok(Device::Cpu)
+    }
+}
+
+#[derive(Clone)]
+pub(crate) struct LlamaModels {
+    model: Arc<Llama>,
+    config: Config,
+    cache: Cache,
+    tokenizer: Tokenizer,
+    device: Device,
+}
+
+impl LlamaModels {
+    pub fn new(model_dir: &Path) -> Result<Self> {
+        let tokenizer_path = model_dir.join(TOKENIZER_FILENAME);
+        let config_path = model_dir.join(CONFIG_FILENAME);
+
+        let dtype = candle::DType::F16;
+        let device = auto_device()?;
+
+        let tokenizer = Tokenizer::from_file(tokenizer_path).map_err(|e| anyhow!(e.to_string()))?;
+        let config: LlamaConfig = serde_json::from_slice(&fs::read(config_path)?)?;
+
+        // TODO: flash attention is supposed to minimize memory read and writes - Do we want to turn it on
+        let config = config.into_config(false);
+        let cache = llama::Cache::new(true, dtype, &config, &device)?;
+
+        let safetensor_files = load_safetensors(model_dir, MODEL_SAFETENSORS_INDEX_FILE)?;
+
+        // TODO: Check if there is a safe way to load the model from the file
+        let vb = unsafe { VarBuilder::from_mmaped_safetensors(&safetensor_files, dtype, &device)? };
```

Comment:
I don't believe we can consider this safe to call. From the docs:

> All file-backed memory map constructors are marked unsafe because of the potential for Undefined Behavior (UB) using the map if the underlying file is subsequently modified, in or out of process. Applications must consider the risk and take appropriate precautions when using file-backed maps. Solutions such as file permissions, locks or process-private (e.g. unlinked) files exist but are platform specific and limited.

It is certainly possible for the files to be modified out of band causing undefined behavior. It seems like `VarBuilder::from_buffered_safetensors` should do the trick at the expense of more memory pressure. 

I think we want to solve this issue before merging.

Reply:
I agree that we want to solve this but I am not sure if `from_buffered_safetensors` is a drop in replacement as it seems to take one binary buffer for the safetensors, I am currently unsure on how to go from multiple file to a single buffer. I will look into this.